# Lab Exercises

## Processing Data with Python, Part 1

### Help

If you need help, please contact one of the TAs for the course.

All TAs from this course can answer your questions, so feel free to attend any of the office hours that fit your schedule

Submit your completed Jupyter notebook on Canvas. You don't need to submit any other files, I have them. :D

----

## Exercises

The exercises for this week are varied, but some deal with `chipotle.tsv`, a dataset concerning Chipotle orders. If you run into any trouble, you may find the texts for this section of the course helpful:

* [The Python Data Science Handbook](https://jakevdp.github.io/PythonDataScienceHandbook/), by Jake VanderPlas, or
* [Python for Data Analysis](https://proquest.safaribooksonline.com/9781491957653), by Wes McKinney.

Oh yeah, and let's import pandas for this notebook.

In [3]:
import pandas as pd

----

#### **Exercise 1**

Using the `iloc` and slicing syntax, slice the following dataframe based on the highlighted blocks in the image. 

First, think of the slicing syntax to grab just the rows you want. Then, think of the slicing syntax for the columns you want. 

Put the row slices *before* the comma and the column slices *after* the comma.

In [ ]:
# here's our DataFrame
indexing_example = pd.DataFrame([[1,2,3],[4,5,6],[7,8,9]])
indexing_example

Select the second two columns of the first two rows.  

![first slice exercise](img/indexing1.png)  

In [1]:
indexing_example.iloc[:2, 1:]

NameError: name 'indexing_example' is not defined

Select the third row.

![first slice exercise](img/indexing2.png)

In [ ]:
indexing_example.iloc[2]

Select the first two columns.

![first slice exercise](img/indexing3.png)

In [ ]:
indexing_example.iloc[:, :2]

Select the first two columns of the second row.

![first slice exercise](img/indexing4.png)

In [ ]:
indexing_example.iloc[1, :2]


#### **Exercise 2**

You're going to be working with Chipotle data from here on out. The data is in a file called `chipotle.tsv`. Since this is a tab-separated file (not a CSV, which uses commas), you'll have to specify `sep='\t'` on your read function.

Import your dataset from `chipotle.tsv` and assign it to a variable called `chip`. Then, view the first 10 entries in your new DataFrame.

In [2]:
import pandas as pd
chip = pd.read_csv('chipotle.tsv', sep='\t')
chip.head(10)

order_id  quantity                              item_name  \
0         1         1           Chips and Fresh Tomato Salsa   
1         1         1                                   Izze   
2         1         1                       Nantucket Nectar   
3         1         1  Chips and Tomatillo-Green Chili Salsa   
4         2         2                           Chicken Bowl   
5         3         1                           Chicken Bowl   
6         3         1                          Side of Chips   
7         4         1                          Steak Burrito   
8         4         1                       Steak Soft Tacos   
9         5         1                          Steak Burrito   

                                  choice_description  item_price  
0                                                NaN        2.39  
1                                       [Clementine]        3.39  
2                                            [Apple]        3.39  
3                                                NaN        2.39  
4  [Tomatillo-Red Chili Salsa (Hot), [Black Beans...       16.98  
5  [Fresh Tomato Salsa (Mild), [Rice, Cheese, Sou...       10.98  
6                                                NaN        1.69  
7  [Tomatillo Red Chili Salsa, [Fajita Vegetables...       11.75  
8  [Tomatillo Green Chili Salsa, [Pinto Beans, Ch...        9.25  
9  [Fresh Tomato Salsa, [Rice, Black Beans, Pinto...        9.25

#### **Exercise 3**

Print the total number of observations in the DataFrame. There are a number of ways you can do this, but the value will be how many individual things were ordered at Chipotle.

In [3]:
print(len(chip))



4622


#### **Exercise 4**

Print the number of columns in the DataFrame, and then print the names of all columns in the DataFrame. This will give you a guide to how the DataFrame is indexed.

In [4]:
print(len(chip.columns))
print(chip.columns)


5
Index(['order_id', 'quantity', 'item_name', 'choice_description',
       'item_price'],
      dtype='object')


#### **Exercise 5**

Calculate, for each order number, how much the order costs. 

In other words, you're going to iterate through the DataFrame and, for each order number, add up the `item_price` for each item in the order.

Store the order prices in a list. Then, tell me how much the 100th, 500th, and 1000th order cost. 

In [5]:
chip['item_price'] = chip['item_price'].replace('[\$,]', '', regex=True).astype(float)
order_totals = chip.groupby('order_id')['item_price'].sum().reset_index()
print(order_totals[order_totals['order_id'] == 100]['item_price'].values[0])
print(order_totals[order_totals['order_id'] == 500]['item_price'].values[0])
print(order_totals[order_totals['order_id'] == 1000]['item_price'].values[0])



10.08
12.15
20.5


(10.08, 12.15, 20.5)

#### **Exercise 6**

This assignment is a bit tricky! Check out the Python for Everyone's section on [dictionaries](https://www.py4e.com/html3/09-dictionaries) if you're confused.

Make a new dictionary. Then, iterate through the DataFrame. For each `item_name`, do the following:

* if `dict[item_name]` exists, add `quantity` to it;
* if not, set `dict[item_name]` to `quantity`;

You're essentially counting the number of times that a given item has been ordered.

In [6]:
item_counts = {}
for index, row in chip.iterrows():
    item_name = row['item_name']
    quantity = row['quantity']
    if item_name in item_counts:
        item_counts[item_name] += quantity
    else:
        item_counts[item_name] = quantity
print(item_counts)


{'Chips and Fresh Tomato Salsa': 130, 'Izze': 20, 'Nantucket Nectar': 29, 'Chips and Tomatillo-Green Chili Salsa': 33, 'Chicken Bowl': 761, 'Side of Chips': 110, 'Steak Burrito': 386, 'Steak Soft Tacos': 56, 'Chips and Guacamole': 506, 'Chicken Crispy Tacos': 50, 'Chicken Soft Tacos': 120, 'Chicken Burrito': 591, 'Canned Soda': 126, 'Barbacoa Burrito': 91, 'Carnitas Burrito': 60, 'Carnitas Bowl': 71, 'Bottled Water': 211, 'Chips and Tomatillo Green Chili Salsa': 45, 'Barbacoa Bowl': 66, 'Chips': 230, 'Chicken Salad Bowl': 123, 'Steak Bowl': 221, 'Barbacoa Soft Tacos': 25, 'Veggie Burrito': 97, 'Veggie Bowl': 87, 'Steak Crispy Tacos': 36, 'Chips and Tomatillo Red Chili Salsa': 50, 'Barbacoa Crispy Tacos': 12, 'Veggie Salad Bowl': 18, 'Chips and Roasted Chili-Corn Salsa': 18, 'Chips and Roasted Chili Corn Salsa': 23, 'Carnitas Soft Tacos': 40, 'Chicken Salad': 9, 'Canned Soft Drink': 351, 'Steak Salad Bowl': 31, '6 Pack Soft Drink': 55, 'Chips and Tomatillo-Red Chili Salsa': 25, 'Bowl': 